scripts/reactor_deployment.py

```diff
@@ -0,0 +1,496 @@
+    # replacement reactor that is new that year, but not new overall because it
+    # is replacing itself).
+    df = num_react_to_cap(df, ar_dict)
+
+    return df
+
+
+def rand_deployment(df, base_col, ar_dict,
+                    set_seed=False, rough=True, tolerance=5):
+    """
+    This function randomly deploys reactors from the dictionary of
+    reactors to meet a capacity need.
+
+    There are two implementations:
+    1) rough (rough=True): if a reactor greater than the remaining
+    capacity is proposed, the deployment ends.
+    2) [IN-PROGRESS] complete (rough=False): the algorithm keeps
+    trying to deploy randomly selected reactors until the capacity demand
+    has been met.
+
+    Parameters
+    ----------
+    df: pandas dataframe
+        The dataframe of capacity information.
+    base_col: str
+        The string name corresponding to the column of capacity that the
+        algorithm is deploying reactors to meet.
+    ar_dict: dictionary
+        A dictionary of reactors with information of the form:
+        {reactor: [Power (MWe), capacity_factor (%), lifetime (yr)]}.
+    set_seed: bool
+        A True/False value that determines whether the seed used for the random
+        number is set or varies based on time.
+    rough: bool
+        A True/False value that determines whether the initial deployment is
+        rough or complete.
+    tolerance: float/int
+        The capacity tolerance to which reactors are deployed in the complete
+        deployment case (i.e., when rough=False). Without this, convergence
+        is tricky to achieve.
+    """
+
+    # initialize the number of reactor columns
+    for reactor in ar_dict.keys():
+        if f'num_{reactor}' not in df:
+            df[f'num_{reactor}'] = 0
+        else:
+            pass
+
+    for year in range(len(df[base_col])):
+        years_capacity = df[base_col][year]
+        # I set the limit this way so that something close to 0 could still
+        # deploy the smallest reactor.
+
+        if set_seed is False:
+            # sample random number based on time
+            c = datetime.now()
+            real_seed = int(c.strftime('%y%m%d%H%M%S'))
+        else:
+            real_seed = 20240527121205
+
+        rng = np.random.default_rng(seed=real_seed)
+
+        while years_capacity > -(ar_dict[list(ar_dict.keys())[-1]][0] + 1):
+            rand_reactor = rng.integers(0, len(ar_dict.keys()))
+
+            # identify random reactor
+            deployed = list(ar_dict.keys())[rand_reactor]
+
+            if ar_dict[deployed][0] > years_capacity:
+                if rough is True:
+                    # for a much rougher check, use break
+                    break
+                elif rough is False:
+                    # for a more accurate, but much much longer run
+                    # todo, finish this ensuring it can converge
+                    print('This feature is unstable.')
```

Comment:
Does it print this multiple times? Might be better to use `Raise NotImplementedError` if you don't want it used yet

Reply:
Oouu, I didn't think about that. The problem is that the feature could theoretically work, there's just no guarantee that it will converge. I've switched to an error.